cpp/include/cudf/io/text/data_chunk_source.hpp

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cudf/utilities/span.hpp>
+
+#include <rmm/cuda_stream_pool.hpp>
+#include <rmm/device_buffer.hpp>
+
+namespace cudf {
+namespace io {
+namespace text {
+
+/**
+ * @brief represents a possibly-shared view over device memory.
+ */
+struct data_chunk {
+  data_chunk(device_span<char const> data) : _data(data) {}
+
+  operator cudf::device_span<char const>() { return _data; }
+
+  uint32_t size() const { return _data.size(); }
+
+ private:
+  device_span<char const> _data;
+};
+
+/**
+ * @brief a reader capable of producing views over device memory
+ *
+ */
+class data_chunk_reader {
```

Comment:
I think this needs some more documentation. At minimum, what does `size` mean here? What kind of data is it returning a view of?

Reply:
`size` is the number of bytes/chars the caller wants. `data_chunk` will contain a possibly _different_ size that indicates how many bytes/chars the consumer actually gets access to (for instance, if asking for 100 bytes at eof will results in 0 bytes).

I'll update documentation to reflect this.